File: incubator/viewers/graphql/viewer/src/main/java/org/apache/causeway/viewer/graphql/viewer/source/TypeMapper.java
/*
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
package org.apache.causeway.viewer.graphql.viewer.source;

import java.math.BigDecimal;
import java.math.BigInteger;
import java.util.Arrays;
import java.util.List;

import org.apache.causeway.core.metamodel.facets.actcoll.typeof.TypeOfFacet;
import org.apache.causeway.core.metamodel.spec.ObjectSpecification;
import org.apache.causeway.core.metamodel.spec.feature.ObjectAction;
import org.apache.causeway.core.metamodel.spec.feature.ObjectActionParameter;

import graphql.Scalars;
import graphql.schema.GraphQLInputType;
import graphql.schema.GraphQLList;
import graphql.schema.GraphQLType;
import graphql.schema.GraphQLTypeReference;

public class TypeMapper {

    private static List<Class<?>> mapToInteger = Arrays.asList(
            int.class, Integer.class, Short.class, short.class, BigInteger.class);
    private static List<Class<?>> mapToLong = Arrays.asList(Long.class, long.class, BigDecimal.class);
    private static List<Class<?>> mapToBoolean = Arrays.asList(Boolean.class, boolean.class);

    public static GraphQLType typeFor(final Class<?> c){
        if (mapToInteger.contains(c)){
            return Scalars.GraphQLInt;
        }
        if (mapToLong.contains(c)){
            return Scalars.GraphQLFloat;
        }
        if (mapToBoolean.contains(c)){
            return Scalars.GraphQLBoolean;
        }
        return Scalars.GraphQLString;
    }

    public static GraphQLInputType inputTypeFor(final ObjectActionParameter objectActionParameter){
        ObjectSpecification elementType = objectActionParameter.getElementType();
        switch (elementType.getBeanSort()) {
            case ABSTRACT:
            case ENTITY:
            case VIEW_MODEL:

                return GraphQLTypeReference.typeRef(_Utils.GQL_INPUTTYPE_PREFIX + _Utils.logicalTypeNameSanitized(elementType.getLogicalTypeName()));

            case VALUE:
                return (GraphQLInputType) typeFor(elementType.getCorrespondingClass());

            case COLLECTION:
                // TODO ...
            default:
                // for now
                return Scalars.GraphQLString;
        }

    }

    public static GraphQLType typeForObjectAction(final ObjectAction objectAction){
        ObjectSpecification objectSpecification = objectAction.getReturnType();
        switch (objectSpecification.getBeanSort()){

            case COLLECTION:

                TypeOfFacet facet = objectAction.getFacet(TypeOfFacet.class);
                if (facet == null) return GraphQLList.list(Scalars.GraphQLString); // TODO: for now ... Investigate why this can happen
                ObjectSpecification objectSpecificationForElementWhenCollection = facet.elementSpec();
                return GraphQLList.list(outputTypeFor(objectSpecificationForElementWhenCollection));

            case VALUE:
            case ENTITY:
            case VIEW_MODEL:
            default:
                return outputTypeFor(objectSpecification);

        }
    }

    public static GraphQLType outputTypeFor(final ObjectSpecification objectSpecification){

        switch (objectSpecification.getBeanSort()){
            case ABSTRACT:
            case ENTITY:
            case VIEW_MODEL:
                return GraphQLTypeReference.typeRef(_Utils.logicalTypeNameSanitized(objectSpecification.getLogicalTypeName()));

            case VALUE:
                return typeFor(objectSpecification.getCorrespondingClass());

            case COLLECTION:
                // should be noop
                return null;

            default:
                // for now
                return Scalars.GraphQLString;
        }
    }


}

File: incubator/clients/kroviz/src/main/kotlin/org/apache/causeway/client/kroviz/to/TypeMapper.kt
/*
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
package org.apache.causeway.client.kroviz.to

import org.apache.causeway.client.kroviz.utils.XmlHelper

enum class ValueType(val type: String) {
    BUTTON("Button"),
    BOOLEAN("Boolean"),
    DATE("Date"),
    HTML("Html"),
    IFRAME("IFrame"),
    IMAGE("Image"),
    NUMERIC("Numeric"),
    PASSWORD("Password"),
    TEXT("Text"),
    TEXT_AREA("TextArea"),
    TIME("Time"),
    SIMPLE_SELECT("SimpleSelect"),
    SLIDER("Slider"),
    SHELL("Shell"),
    SVG_MAPPED("Map"),
    SVG_INLINE("Inline"),
    VEGA("Vega"),
}

class TypeMapper {

    fun match(member: Member): String {
        val mf = member.format
        val mex = member.extensions?.xCausewayFormat
        return when {
            mf == "int" -> ValueType.NUMERIC.type
            mf == "date" -> ValueType.DATE.type
            mf == "date-time" -> ValueType.TIME.type
            mf == "utc-millisec" -> ValueType.TIME.type
            mex == "boolean" -> ValueType.BOOLEAN.type
            //mex == "javasqltimestamp" -> TypeMapperType.DATE.type
            //mex == "javautildate" -> TypeMapperType.DATE.type
            else -> {
                match(member.value)
            }
        }
    }

    private val isoDate = Regex("^([+-]?\\d{4}(?!\\d{2}\\b))((-?)((0[1-9]|1[0-2])(\\3([12]\\d|0[1-9]|3[01]))?|W([0-4]\\d|5[0-2])(-?[1-7])?|(00[1-9]|0[1-9]\\d|[12]\\d{2}|3([0-5]\\d|6[1-6])))([T\\s]((([01]\\d|2[0-3])((:?)[0-5]\\d)?|24:?00)([.,]\\d+(?!:))?)?(\\17[0-5]\\d([.,]\\d+)?)?([zZ]|([+-])([01]\\d|2[0-3]):?([0-5]\\d)?)?)?)?\$")

    private fun match(value: Value?): String {
        val contentStr = value?.content.toString()
        return when {
            isoDate.matches(contentStr) -> ValueType.DATE.type
            XmlHelper.isXml(contentStr) -> ValueType.HTML.type
            contentStr.startsWith(":Notice:") -> ValueType.HTML.type
            contentStr.startsWith("link:") -> ValueType.HTML.type
            else -> ValueType.TEXT.type
        }
    }

    fun forType(type: String): ValueType {
        console.log("[TM_forType] $type")
        when (type) {
            ValueType.BOOLEAN.type -> return ValueType.BOOLEAN
            ValueType.DATE.type -> return ValueType.DATE
            ValueType.HTML.type -> return ValueType.HTML
            ValueType.IFRAME.type -> return ValueType.IFRAME
            ValueType.IMAGE.type -> return ValueType.IMAGE
            ValueType.NUMERIC.type -> return ValueType.NUMERIC
            ValueType.PASSWORD.type -> return ValueType.PASSWORD
            ValueType.TEXT.type -> return ValueType.TEXT
            ValueType.TEXT_AREA.type -> return ValueType.TEXT_AREA
            ValueType.TIME.type -> return ValueType.TIME
            ValueType.SIMPLE_SELECT.type -> return ValueType.SIMPLE_SELECT
            ValueType.SLIDER.type -> return ValueType.SLIDER
            ValueType.SVG_MAPPED.type -> return ValueType.SVG_MAPPED
            ValueType.SVG_INLINE.type -> return ValueType.SVG_INLINE
            ValueType.VEGA.type -> return ValueType.VEGA
            else -> {
                return ValueType.TEXT
            }
        }
    }

}
